### DISCUSSION TITLE
Fast Refresh Performing Full Reload Issue in Next.js v14 Only

### DISCUSSION THREAD

**Discussion by Choi-Jiwon-38** [vercel/next.js#69600] [Help] [answered]
### Summary

Hello,

I've encountered a serious issue after upgrading to Next.js v14. In all previous versions, Fast Refresh worked perfectly, but after upgrading to v14, it now causes an infinite reload loop whenever I make changes during development.

This issue only occurs in v14; everything works as expected in earlier versions. The infinite reload not only disrupts the development process but also makes it difficult to test changes effectively.

Could you please advise if there are any known solutions or patches for this issue? I would appreciate any guidance or troubleshooting steps.

Thank you for your support.

![image](https://github.com/user-attachments/assets/1ce898c5-49b1-44c3-a282-a3c45ace3bba)

**Comment by Juneezee** [comment]
Do you have a client component that is `async`? See https://github.com/vercel/next.js/issues/49881#issuecomment-2269353762

**Reply by Choi-Jiwon-38** [reply]
Upon checking, there are no client components inside the async function. Additionally, clearing all browser caches resolves the issue, and it works fine in the production environment. However, the problem only occurs when testing locally.

**Reply by Juneezee** [reply]
> clearing all browser caches resolves the issue, and it works fine in the production environment

Have you tried deleting the `.next` directory and run `next dev` again? It does seem like a cache issue based on your description

> clearing all browser caches

You could also try doing a Hard Reload on the page with <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>R</kbd> (https://developer.chrome.com/docs/devtools/shortcuts) and see if it works before clearing all browser caches.

**Reply by Choi-Jiwon-38** [reply]
Yes, using `Ctrl` + `Shift` + `R` can stop the infinite reload, but I want to prevent the infinite reload from occurring in the first place.

One difference I've noticed is that in another project with version 14, files like `_app.js` and `main.js` are not cached in memory, whereas in the project experiencing the infinite reload, they are being cached. Could this be the cause of the issue?

<img width="706" alt="image" src="https://github.com/user-attachments/assets/a30819d3-4acb-43b6-938c-5a9c5845d456">

**Reply by Juneezee** [reply]
> Yes, using <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>R</kbd> can stop the infinite reload

At this point this looks most likely a cache problem. Are you using App Router or Pages Router?

You said this issue only occurs after you upgraded to Next.js 14. It's hard to tell whether it is a bug on Next.js side or an incompatibility issue with one of your project dependencies.

If you have some extra time, you could try creating a fresh Next.js 14 project with `create-next-app` and move your code one by one to produce a minimal reproducible example that can reproduce the bug.

**Reply by Choi-Jiwon-38** [reply]
Oh, thank you for confirming :) As you mentioned, the cache issue was a major factor.

```
const nextConfig = {
  ...
  headers: async () => {
      return [
         ...
        {
          source: '/:path*(svg|jpg|jpeg|png|gif|ico|webp)',
          headers: [{ key: 'Cache-Control', value: 'public, max-age=31536000, must-revalidate' }],
        },
        ...
      ];
    },
  ...
}
```

The main cause was that in `next.config.js`, the header configuration for `.js` files was set to cache them.

However, in Next.js version 13, files with `no-cache, private` settings were changed to `public, max-age=31536000, must-revalidate` in version 14.

Was there any change in the development server regarding updates when upgrading versions? The problem is resolved, but I am curious about why there is a difference in the development environment with the same code just due to the version change.

**Reply by Juneezee** [reply]
> The main cause was that in `next.config.js`, the header configuration for `.js` files was set to cache them.

I don't think the default `next.config.js` sets the header for `.js` file to cache them. Did you set them manually?

According to the docs (https://nextjs.org/docs/app/api-reference/next-config-js/headers#cache-control):

> Next.js sets the `Cache-Control` header of `public, max-age=31536000, immutable` for **truly immutable assets**. It cannot be overridden. These immutable files contain a SHA-hash in the file name, so they can be safely cached indefinitely. For example, [Static Image Imports](https://nextjs.org/docs/app/building-your-application/optimizing/images#local-images). You cannot set `Cache-Control` headers in `next.config.js` for these assets.

So in your case https://github.com/vercel/next.js/discussions/69600#discussioncomment-10539254, if you modify your `next.config.js` to cache all `*.js` files, then files without those SHA-hash (e.g. `_app.js`) are cached too, which is not what we wanted.

**Reply by Choi-Jiwon-38** [reply]
Yes, that's correct. In `next.config.js`, `.js` files are configured not to be cached by default. This setting was manually configured.

However, before upgrading to version 14, these settings were only applied in the production environment. In the local development environment, the headers were set to `no-cache, private`. In the production environment, SHA hashes are actually appended (e.g., `_app-{HASH}.js`), and caching works properly.

Are there any notable changes related to the upgrade from version 13 to 14 that might be relevant to this issue? I've been searching but haven't been able to find anything. While I've resolved the problem, I'm curious about the root cause. :(

**Reply by Juneezee** [reply, marked as answer]
> Are there any notable changes related to the upgrade from version 13 to 14 that might be relevant to this issue?

A lot have changes since then. I also couldn't find the release blogs https://nextjs.org/blog/next-14, https://nextjs.org/blog/next-14-1, https://nextjs.org/blog/next-14-2 mentioning anything regarding `Cache-Control`.

Digging through the Git history, I found this PR that updates the `Cache-Control` section of the docs: https://github.com/vercel/next.js/pull/68092

Prior to the PR, the docs was

```markdown
You cannot set `Cache-Control` headers in `next.config.js` for pages or assets, as these headers will be overwritten in production to ensure that responses and static assets are cached effectively.
```

And the closest code that I can find is

https://github.com/vercel/next.js/blob/f90c6d71472d3dbc21bdd30cb7fbc88feeadde0f/packages/next/src/server/lib/router-server.ts#L433-L445

So it looks like yours configuration is overriding the `'no-store, must-revalidate'` in development mode. Looks like a bug to me but I'm not sure too

**Reply by Choi-Jiwon-38** [reply]
Thank you for confirming. The information you provided was very helpful in resolving the issue I encountered.

The infinite reload problem I experienced was related to caching, and it seems that the behavior of the development mode override in my project changed due to a version update.

If you come across any further insights or solutions to similar issues in the future, I would greatly appreciate it if you could share them. :)
